### DISCUSSION TITLE
System.FormatException: Input string was not in a correct format.

### DISCUSSION THREAD

**Discussion by ArtemHachaturov** [DaveSkender/Stock.Indicators#1194] [Help and support]
Hello, please help me, I’ve been suffering for two days now, I have a problem with converting data to a decimal fraction, float, and specifically with

import pandas as pd
from stock_indicators import Quote

# Read the CSV file
df = pd.read_csv('./quotes.csv', parse_dates=['Time'])

quotes = [ 
  Quote(date, open, high, low, close, volume) 
    for date, open, high, low, close, volume 
    in zip(df['Time'],
           df['Open'], 
           df['High'], 
           df['Low'],
           df['Close'], 
           df['Volume'], strict=True)]

print(quotes)

------------------------------------------

Traceback (most recent call last):
  File "C:\Users\russt\AppData\Local\Programs\Python\Python311\Lib\site-packages\stock_indicators\_cstypes\decimal.py", line 22, in __new__
    return CsDecimal.Parse(str(decimal))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
System.FormatException: Input string was not in a correct format.
   at System.Number.ThrowOverflowOrFormatException(ParsingStatus status, TypeCode type) 
   at System.Decimal.Parse(String s)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "e:\Binance Bot\python\13.py", line 8, in <module>
    quotes = [
             ^
  File "e:\Binance Bot\python\13.py", line 9, in <listcomp>
    Quote(date, open, high, low, close, volume)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\russt\AppData\Local\Programs\Python\Python311\Lib\site-packages\stock_indicators\indicators\common\quote.py", line 61, in __init__
    self.open: Decimal = open if open else 0
    ^^^^^^^^^
  File "C:\Users\russt\AppData\Local\Programs\Python\Python311\Lib\site-packages\stock_indicators\indicators\common\quote.py", line 25, in _set_open
    quote.Open = CsDecimal(value)
                 ^^^^^^^^^^^^^^^^
  File "C:\Users\russt\AppData\Local\Programs\Python\Python311\Lib\site-packages\stock_indicators\_cstypes\decimal.py", line 24, in __new__
    raise ValueError("You may be using numeric data that is incompatible with your locale environment settings.\n"
ValueError: You may be using numeric data that is incompatible with your locale environment settings.
For example, you may be using decimal points instead of commas.

[quotes.csv](https://github.com/DaveSkender/Stock.Indicators/files/15367062/quotes.csv)

**Comment by DaveSkender** [comment]
Let us know if these conversations helps resolve your issue:

- #1165
- #1164

**Reply by ArtemHachaturov** [reply]
Thank you very much for the information provided, after reading it it became presumably clear that my current_locale = locale.getdefaultlocale()[0] "ru-Ru" is not compatible with data that looks correct in appearance but is not compatible with quote.High = CsDecimal(value). for example "223.123355" which are used by df = pd.read_csv('quotes.csv', parse_dates=['Time']) which is visible in the output

print(df.dtypes)

# Time      datetime64[ns]
# Open             float64
# High             float64
# Low              float64
# Close            float64
# Volume         float64


however, I also can't create a Quote object even when using locale.setlocale(locale.LC_NUMERIC, 'en_US.UTF-8') or locale.setlocale(locale.LC_ALL, 'C').

please suggest a solution for converting the Quote object

**Reply by ArtemHachaturov** [reply]
from stock_indicators import Quote
from decimal import Decimal

import locale
locale.setlocale(locale.LC_NUMERIC, 'en_US.UTF-8')


date = pd.Timestamp('2024-05-19')
open_price = 123.45
high_price = 145.67
low_price = 100.00
close_price = 135.79
volume = 100000


open_decimal = Decimal(str(open_price))
high_decimal = Decimal(str(high_price))
low_decimal = Decimal(str(low_price))
close_decimal = Decimal(str(close_price))

print(type(open_decimal))

quote = Quote(date, open_decimal, high_decimal, low_decimal, close_decimal, volume)

print(quote)

**Reply by ArtemHachaturov** [reply]
quote.Open = CsDecimal(value)
                 ^^^^^^^^^^^^^^^^
  File "C:\Users\russt\AppData\Local\Programs\Python\Python311\Lib\site-packages\stock_indicators\_cstypes\decimal.py", line 24, in __new__
    raise ValueError("You may be using numeric data that is incompatible with your locale environment settings.\n"
ValueError: You may be using numeric data that is incompatible with your locale environment settings.
For example, you may be using decimal points instead of commas.

**Reply by DaveSkender** [reply]
check for `pd` date/timestamp compatibility too, you may need to set locale category to `locale.LC_ALL`, or convert your dates to a compatible string format.

**Reply by DaveSkender** [reply]
does this work okay?
open_price = 123,45
open_decimal = Decimal(str(open_price))
print(type(open_decimal))

**Reply by ArtemHachaturov** [reply]
This method df = pd.read_csv('quotes.csv', parse_dates=['Time']) correctly executes parse_dates=['Time'] without any obstacles and it passes through perfectly Time datetime64[ns] there is no problem with this.

But now there is a problem with the data and locale.LC_ALL does not help (locale.LC_NUMERIC, 'en_US.UTF-8') still the same error

**Reply by ArtemHachaturov** [reply]
open_price = 123.45
open_decimal = Decimal(str(open_price))
print(type(open_decimal))     #<class 'decimal.Decimal'>

#but if

quote = Quote(date, open_decimal, 0, 0, 0, 0)

------------------

Traceback (most recent call last):
  File "C:\Users\russt\AppData\Local\Programs\Python\Python311\Lib\site-packages\stock_indicators\_cstypes\decimal.py", line 22, in __new__
    return CsDecimal.Parse(str(decimal))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
System.FormatException: Input string was not in a correct format.   at System.Number.ThrowOverflowOrFormatException(ParsingStatus 
status, TypeCode type)
   at System.Decimal.Parse(String s)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "e:\Binance Bot\python\12.py", line 51, in <module>       
    quote = Quote(date, open_decimal, 0, 0, 0, 0)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\russt\AppData\Local\Programs\Python\Python311\Lib\site-packages\stock_indicators\indicators\common\quote.py", line 61, in __init__
    self.open: Decimal = open if open else 0
    ^^^^^^^^^
  File "C:\Users\russt\AppData\Local\Programs\Python\Python311\Lib\site-packages\stock_indicators\indicators\common\quote.py", line 25, in _set_open
    quote.Open = CsDecimal(value)
                 ^^^^^^^^^^^^^^^^
  File "C:\Users\russt\AppData\Local\Programs\Python\Python311\Lib\site-packages\stock_indicators\_cstypes\decimal.py", line 24, in __new__
    raise ValueError("You may be using numeric data that is incompatible with your locale environment settings.\n"
ValueError: You may be using numeric data that is incompatible with your locale environment settings.
For example, you may be using decimal points instead of commas.  
PS E:\Binance Bot\python>

**Reply by ArtemHachaturov** [reply]
import locale
locale.setlocale(locale.LC_ALL, 'en_US.UTF-8')

Does not help

**Comment by DaveSkender** [comment]
I think we're going to need to setup a rather elaborate set of tests using different local settings on a GitHub Actions runner to better understand exactly what's happening.  It's difficult to test on a local computer, and I'm not 100% confident I know what's happening.

Any one of these can be a problem if they're not aligned on locale for date/time and numerics:

- source quotes format
- environment/machine locale
- dotnet native locale
- python native locale
- Python.NET locale handling
- pandas dataFrame locale handling
- DLL compiled _default_ locale handling

cc @LeeDongGeon1996

**Reply by DaveSkender** [reply]
I've created a tracking issue:

- https://github.com/facioquo/stock-indicators-python/issues/382
